distributed/scheduler.py

```diff
@@ -2014,11 +2019,29 @@ def transition_no_worker_processing(self, key, stimulus_id):
                 assert not ts.actor, f"Actors can't be in `no-worker`: {ts}"
                 assert ts in self.unrunnable
 
-            if ws := self.decide_worker_non_rootish(ts):
+            decide_worker = (
+                self.decide_worker_rootish_queuing_disabled
+                if self.is_rootish(ts)
+                else self.decide_worker_non_rootish
+            )
```

Comment:
One of the biggest problems I have right now with the queuing/rootish scheduling is that we have three different `decide_*` functions. From an (internal) API perspective, I don't want to have the burden of making the correct decision about which one of these APIs to call in which circumstance. I just want to call a single `decide_worker`, provide it with sufficient context and it should return the proper worker. Wouldn't this already avoid the problem?

Naively I would expect that this new decide_worker would look approximately like the block in `transition_waiting_processing` https://github.com/dask/distributed/blob/17156e9c2a57609846270141f441d3700a729465/distributed/scheduler.py#L2230-L2242

Isn't this always the correct logic when deciding on a worker?

Reply:
Yes, and this is how I'd originally implemented it in https://github.com/dask/distributed/pull/6614. But based on your feedback (which I agree with) we split it into multiple `decide_worker_*` functions for different cases.

The reason we didn't wrap the three `decide_worker_*` cases into one overall `decide_worker` function, which always "does the right thing", is that the recommendation you make—`no-worker` vs `queued`—changes depending on which function you use.

So then this `decide_worker` function would have to take and mutate a `recommendations` dict, or at least return somehow what recommendation to make. I thought we'd decided this was a pattern we wanted to avoid.

Moreover, we'd then have to implement a `no-worker->queued` and `queued->no-worker` transition. That's not hard, just more complexity. If we don't do https://github.com/dask/distributed/pull/7262, it's maybe the right thing to do instead of this PR.